File: src/test_runner1.js
/**
 * Standalone test script, running direct function testing to verify core
 * functionalities independently and sequentially as minimal regression.
 * • Built-in test mode with mock data from config.json
 * • Online mode returns live API calls from external services
 * • Runnable in both browser console and Node.js CLI
 *
 * Browser usage:
 *   import('./src/test_runner.js').then(m => m.testRunner());
 *
 * Node.js CLI usage:
 *   node src/test_runner.js [--test-mode (default) |--online]
 */

// Environment detection
const isBrowser = typeof window !== 'undefined';
const isNode = typeof process !== 'undefined' && process.versions?.node;

// Mock functions for Node.js environment
if (isNode) {
  global.window = {
    location: { hostname: 'localhost', search: '?test=true' },
    TEST_MODE: true,
    APP_CONFIG: {},
  };
  const store = {};
  window.localStorage = {
    getItem: (k) => (k in store ? store[k] : null),
    setItem: (k, v) => {
      store[k] = String(v);
    },
    removeItem: (k) => {
      delete store[k];
    },
    clear: () => {
      for (const k in store) delete store[k];
    },
  };
  global.localStorage = window.localStorage;
  global.document = {
    querySelector: () => null,
    getElementById: () => null,
    readyState: 'complete',
    addEventListener: () => {},
  };
}

// Simple logging
const log = (...args) => console.log(...args);
function error(message) {
  console.error(`❌ ${message}`);
}

// Configuration loading
async function loadConfig() {
  let config = null;
  try {
    if (isBrowser) {
      // Browser environment - call getConfig
      const { getConfig } = await import('./utils.js');
      config = await getConfig();
    } else if (isNode) {
      // Node.js environment
      window.APP_CONFIG.GOOGLE_MAPS_API_KEY =
        process.env.GOOGLE_MAPS_API_KEY || null;
      window.APP_CONFIG.OPENAI_API_KEY = process.env.OPENAI_API_KEY || null;
      config = {
        defaults: {
          default_location: {
            name: 'San Francisco',
            country: 'United States',
            country_code: 'US',
            lat: 37.7749,
            lon: -122.4194,
          },
          search_radius: 15,
        },
        test_mode: {
          test_landmarks: [
            {
              name: 'Golden Gate Bridge',
              lat: 37.8199,
              lon: -122.4783,
              loc: 'San Francisco',
            },
          ],
        },
      };
    }
    return config;
  } catch (error) {
    log(`⚠️ Failed to load configuration: ${error.message}`);
    return null;
  }
}

import { testI18n } from './lion.js';

async function runAllTests() {
  log('✅ I18n passed', testI18n());
  log('Testing getConfig()');
  const config = await loadConfig();
  if (
    !config?.defaults ||
    !config?.test_mode ||
    !config?.test_mode?.test_landmarks ||
    !config.test_mode.test_landmarks?.length
  ) {
    error('getConfig returned invalid config');
    return false;
  }
  log('✅ getConfig passed', { hasDefaults: !!config.defaults });

  // Check default location
  const default_location = config?.defaults?.default_location;
  const default_lat = default_location.lat;
  const default_lon = default_location.lon;
  log(`🌍 Default location: (${default_lat}, ${default_lon})`);
  const default_radius = config?.defaults?.search_radius;

  log('Testing getLocationDetails()');
  const { getLocationDetails } = await import('./gmap.js');
  const locationData = await getLocationDetails(default_lat, default_lon);
  if (
    !locationData?.locationName ||
    locationData.locationName.toLowerCase().includes('unknown')
  ) {
    error('getLocationDetails returned invalid data');
    return false;
  }
  log('✅ getLocationDetails passed', locationData.locationName);

  log('Testing getLocationCoord()');
  const { getLocationCoord } = await import('./gmap.js');
  const coords = await getLocationCoord(default_location.name);
  if (!coords) {
    error('getLocationCoord returned invalid data');
    return false;
  }
  log('✅ getLocationCoord passed', coords.lat, coords.lon);

  log('Testing getLandmarksWithGPT()');
  const { getLandmarksWithGPT } = await import('./openai.js');
  const landmarkData = await getLandmarksWithGPT(
    locationData,
    default_lat,
    default_lon
  );
  if (
    !landmarkData ||
    !landmarkData.landmarks ||
    !Array.isArray(landmarkData.landmarks)
  ) {
    error('getLandmarksWithGPT returned invalid landmarks');
    return false;
  }
  log('✅ getLandmarksWithGPT passed', {
    count: landmarkData.landmarks.length,
  });

  log('Testing setCachedLandmarks()');
  const { setCachedLandmarks } = await import('./cache.js');
  setCachedLandmarks(default_lat, default_lon, default_radius, landmarkData);
  log('✅ setCachedLandmarks passed');

  log('Testing getCachedLandmarks()');
  const { getCachedLandmarks } = await import('./cache.js');
  const cachedData = getCachedLandmarks(
    default_lat,
    default_lon,
    default_radius
  );
  if (!cachedData || !cachedData.landmarks) {
    error('getCachedLandmarks failed to retrieve data');
    return false;
  }
  log('✅ getCachedLandmarks passed', {
    count: cachedData.landmarks.length,
  });

  log('Testing get_landmark_data()');
  const { landmarkService } = await import('./backend.js');
  const backendData = await landmarkService.get_landmark_data(
    default_lat,
    default_lon
  );
  if (!backendData || !backendData.landmarks) {
    error('get_landmark_data returned invalid data');
    return false;
  }
  log('✅ get_landmark_data passed', { count: backendData.landmarks.length });

  log('Testing getWikiImageURL()');
  const { getWikiImageURL } = await import('./services.js');
  const imageUrl = await getWikiImageURL(
    config?.test_mode?.test_landmarks[0].name
  );
  if (typeof imageUrl !== 'string') {
    error('getWikiImageURL returned invalid type');
    return false;
  }
  log('✅ getWikiImageURL passed', { hasUrl: imageUrl.length > 0 });

  log('Testing PlaceTextSearch()');
  const { PlaceTextSearch } = await import('./gmap.js');
  const searchResult = await PlaceTextSearch(default_location.name);
  if (!searchResult || !searchResult.landmarks) {
    error('PlaceTextSearch returned invalid data');
    return false;
  }
  log('✅ PlaceTextSearch passed', { count: searchResult.landmarks.length });

  log('Testing queryLocationWithGPT()');
  const { queryLocationWithGPT } = await import('./openai.js');
  const queryResult = await queryLocationWithGPT(default_location.name);
  if (!queryResult) {
    error('queryLocationWithGPT returned invalid data');
    return false;
  }
  log('✅ queryLocationWithGPT passed', { count: queryResult.landmarks });

  return true;
}

async function testRunner() {
  const { isTestMode } = await import('./interfaces.js');
  console.log(
    `🧪 Run test enabled - ${
      isTestMode() ? 'test mode (mock data)' : 'online (live API)'
    } on ${isBrowser ? 'Browser' : 'Node'}`
  );

  const success = await runAllTests();
  if (success) console.log('✅ PASS: All tests completed successfully');
  else console.error('❌ FAIL: Some tests failed! Check the logs for details');
}

// CLI argument parsing for Node.js
async function parseArgs() {
  if (!isNode) return;
  const { enableTestMode } = await import('./interfaces.js');
  const args = process.argv.slice(2);
  for (const arg of args) {
    if (arg === '--online') {
      enableTestMode(false);
    } else if (arg === '--test-mode') {
      enableTestMode(true);
    } else if (arg === '--help' || arg === '-h') {
      console.log(`
Standalone test script for minimal regression:
Usage: node src/test_runner.js [options]

Options:
  --test-mode Run in test mode with mock data (default)
  --online    Run in online mode with live API calls
  --help, -h  Show this help message
      `);
      process.exit(0);
    }
  }
}

async function parseURLParams() {
  if (!isBrowser) return;
  const { enableTestMode } = await import('./interfaces.js');
  const urlParams = new URLSearchParams(window.location.search);
  if (urlParams.has('test')) {
    const param = urlParams.get('test');
    if (param === 'false' || param === '0') {
      console.log('Run test disabled - skipping client-side tests');
      return;
    }

    if (!urlParams.has('online')) enableTestMode(true);
    await testRunner();
    enableTestMode(false);
  }
}

export async function main() {
  if (isBrowser) {
    await parseURLParams();
  } else if (isNode) {
    function getFilename(filePathOrUrl) {
      const parts = filePathOrUrl.split(/[/\\]/);
      return parts[parts.length - 1];
    }

    const scriptFilename = getFilename(process.argv[1]);
    const metaFilename = getFilename(import.meta.url);
    // Node.js equivalent of if __name__ == "__main__":
    if (scriptFilename === metaFilename) {
      await parseArgs();
      await testRunner();
    }
  }
}

// Auto-run tests
main();

File: tests/test_runner1.py
#!/usr/bin/env python3
"""
Standalone test script, running direct function testing to verify core
functionalities independently and sequentially as minimal regression.
• Built-in test mode with mock data from config.json
• Online mode returns live API calls from external services

# Test mode enabled: --test-mode (default)
# Test mode enabled: os.environ["TEST_MODE"] = "1"
# Disable Test mode: --online

# Set DEBUG for more detailed output
# Debug log enabled: --debug
# Debug log enabled: os.environ["DEBUG"] = "1"
"""

import os
import sys

# Add the project root directory to the Python path
sys.path.insert(
    0, os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
)

from utils import config_utils
from services.landmark_service import LandmarkService
from services.location_service import LocationService
from services.llm_service import LLMService
from utils import coord_utils
from components.cache_manager import cache_manager
import logging

logger = logging.getLogger("tester1")


def main():
    config_utils.enable_test_mode()
    config_utils.setup_logging()
    logger.info(f"TEST_MODE: {config_utils.is_test_mode_enabled()} ✅")

    # Initialize services and test them
    location_service = LocationService()
    llm_service = LLMService()
    landmark_service = LandmarkService()

    # Load configuration with test fixtures
    config = config_utils.load_config(force_reload=True)
    if not config or "defaults" not in config or "test_mode" not in config:
        logger.error("Failed to load configuration ❌")
        return False

    # Check if test fixtures are present
    test_fixtures = config.get("test_mode", {}).get("test_landmarks", [])
    if len(test_fixtures) == 0:
        logger.error("No test fixtures found in config.json ❌")
        return False
    logger.info("Test fixtures loaded successfully ✅")

    # Check default location
    default_location = config_utils.get_default_location()
    default_lat, default_lon = default_location["lat"], default_location["lon"]
    logger.info(f"🌍 Default location: ({default_lat}, {default_lon}) ✅")
    default_radius = config["defaults"]["search_radius"]

    """
    result = cache_manager.clear_cache()
    if not result:
        logger.error("clear_cache ❌")
        return False
    logger.info("clear_cache ✅")
    """

    # Test location service
    location_data = location_service.get_location_details(
        default_lat, default_lon
    )
    location_name = location_data.get("location_name")
    if not location_name or "unknown" in location_name.lower():
        logger.error("Location service failed to return a location name ❌")
        return False
    logger.info(f"Location service returned: {location_name} ✅")

    coords = location_service.get_location_coord(default_location["name"])
    if not coords or not coord_utils.validate_coords(coords.lat, coords.lon):
        logger.error("Location service failed to return valid coordinates ❌")
        return False
    logger.info(f"Location service returned: {coords.lat}, {coords.lon} ✅")

    # Test LLM service with default location
    landmarks = llm_service.get_landmarks_with_gpt(
        location_data, default_lat, default_lon, default_radius
    )
    if not landmarks:
        logger.error("LLM service failed to return landmarks ❌")
        return False
    logger.info(f"🏛️  LLM service returned {len(landmarks)} landmarks ✅")

    # Run cache manager tests
    cached_ok = cache_manager.cache_landmarks(
        location_name, default_lat, default_lon, default_radius, landmarks
    )
    if not cached_ok:
        logger.error("cache_landmarks failed ❌")
        return False
    logger.info(f"Cached {len(landmarks)} landmarks ✅")
    cached_landmarks = cache_manager.get_cached_landmarks(
        location_name, default_lat, default_lon, default_radius
    )
    if not cached_landmarks or len(cached_landmarks) == 0:
        print("No cached landmarks retrieved ❌")
        return False
    logger.info(f"🔄 Retrieved {len(cached_landmarks)} cached landmarks ✅")

    # Test landmark service
    landmark_data = landmark_service.get_landmark_data(default_lat, default_lon)
    if not landmark_data or not landmark_data.get("landmarks"):
        logger.error("Landmark service failed to return landmark data ❌")
        return False
    logger.info(
        f"📍 Landmark service returned {len(landmark_data.get('landmarks', []))} landmarks"
        + f" for {landmark_data.get('location', 'Unknown')} ✅"
    )

    image_url = landmark_service.get_landmark_image(
        landmark_data["landmarks"][0]["name"]
    )
    if not image_url:
        logger.error("Landmark service failed to return an image URL ❌")
        return False
    logger.info(f"Got image URL: {image_url} ✅")

    loc_data = location_service.text_search_places(default_location["name"])
    if not loc_data:
        logger.error("Text search failed to return location ❌")
        return False
    logger.info(f"🔍 Text search returned location: {loc_data} ✅")

    loc_data = llm_service.query_location_with_gpt(default_location["name"])
    if not loc_data:
        logger.error("LLM service failed to return location ❌")
        return False
    logger.info(f"LLM service returned location: {loc_data} ✅")

    return True


if __name__ == "__main__":
    success = main()
    if not success:
        print("❌ FAIL: Some tests failed! Check the logs for details")
        exit(1)
    else:
        print("✅ PASS: All tests completed successfully")
